ext/msgpack/custom.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * MessagePack for Ruby
+ *
+ * Copyright (C) 2008-2015 Sadayuki Furuhashi, Scott Francis
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "custom.h"
+#include "packer.h"
+#include "packer_class.h"
+#include "core_ext.h"
+
+static ID s_to_msgpack;
+static ID s___to_msgpack;
+static ID s_from_msgpack;
+static ID s_dump;
+static ID s_load;
+static VALUE mMarshal;
+static VALUE ePackError;
+
+static st_table *msgpack_custom_types;
+
+#define MSGPACK_EXT_TYPE_RUBY 1
+
+VALUE msgpack_custom_unpack_type(const char *buffer, size_t sz)
+{
+    const char *p = buffer;
+    size_t data_sz;
+    VALUE str, cls, val;
+
+    if (*p++ != MSGPACK_EXT_TYPE_RUBY) {
+        return Qnil;
+    }
+
+    /* only deserialize types that we've registered */
+    if (!st_lookup(msgpack_custom_types, (st_data_t) p, (st_data_t *) &cls)) {
+        return Qnil;
```

Comment:
That will happen, just not from here. By returning `nil`, the unpacker will return `PRIMITIVE_UNEXPECTED_TYPE` which eventually does raise an error: https://github.com/csfrancis/msgpack-ruby/blob/msgpack_custom_types/ext/msgpack/unpacker.c#L348-L353

I didn't want to raise any errors from here to be consistent with the rest of the unpacker code.